gordon/metrics/log.py

```diff
@@ -0,0 +1,164 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Gordon ships with a default IMetricRelay implementation, which
+outputs metrics to the application logger using the standard
+library `logging` module.
+
+This implementation is used by default if no other IMetricRelay
+implementation is loaded and chosen.
+
+The default configuration that can be overriden is shown below.
+
+.. code-block:: ini
+
+    [metrics-logger]
+    # One of the log levels available from the stdlib logging module.
+    log_level = 'info'
+    # A scaling factor for timing (see: LogTimer)
+    time_unit = 1
+"""
+
+import collections
+import logging
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class LogTimer:
+    """Timer that outputs metrics to the logger.
+
+    Args:
+        metric (dict): Metric to output.
+        logger (.LoggerAdapter): Object used for outputting metrics.
+        time_unit (number): (optional) Scale time unit for use with
+            time.perf_counter(), for example: 1E9 to send nanoseconds.
+    """
+    def __init__(self, metric, logger, time_unit=1):
+        self.metric = metric
+        self.logger = logger
+        self.time_unit = time_unit
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        """Start the timer."""
+        self._start_time = time.perf_counter()
+
+    async def stop(self):
+        """Stop the timer and output the time delta to the logger."""
+        time_elapsed = (
+            time.perf_counter() - self._start_time) * self.time_unit
+        self.metric['value'] = time_elapsed
+        self.logger.log(self.metric)
+
+
+class LoggerAdapter:
+    """Configurable adapter that formats and outputs metrics.
+
+    Args:
+        level (str): Log level at which to output metric.
+    """
+    LOGFMT = '[{metric_name}] value: {value}'
+
+    def __init__(self, level):
+        self.level = level
+        self._logger = logging.getLogger('gordon.metrics-logger')
+
+    def log(self, metric):
+        """Format and output metric.
+
+        Args:
+            metric (dict): Complete metric.
+        """
+        message = self.LOGFMT.format(**metric)
+        if metric['context']:
+            message += ' context: {context}'.format(context=metric['context'])
+        getattr(self._logger, self.level.lower())(message)
```

Comment:
instead of calling `lower()` every message, should we just do it once on initialization?

Reply:
Actually should we call `upper()` on initialization and use [log()](https://docs.python.org/3/library/logging.html#logging.Logger.log) instead?  I think it's a bit clearer than `getattr`.

Reply:
I agree that `lower` should be called once (and maybe `getattr` once on instantiation too), but I'm not sure I'm understanding how that `log` method works?

```pycon
>>> import logging
>>> logger = logging.getLogger('')
>>> logger.log('INFO', 'hi')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lynn/.pyenv/versions/3.6.2/lib/python3.6/logging/__init__.py", line 1368, in log
    raise TypeError("level must be an integer")
TypeError: level must be an integer
>>>
# same with 'info'
```

Reply:
AFAICT it works the same as `info` or `debug`, et. al., but you have to pass the level constant in as the first argument.
```
>>> import logging
>>> logger = logging.getLogger('')
>>> logger.log(logging.INFO, 'hi')
```
So I initially wrote to do `upper()` in the `__init__` so that you could have like 
`self.level = getattr(logging, level.upper())`
and then `level` would be the `logging.LEVEL` rather than a string. 

But then I thought (and edited the comment) why not have the `__init__` take `logging.LEVEL` instead. I guess it's just moving the config -> log level parsing logic from the adapter to the relay, just cuz I thought it fit a little better there, but I could go either way on it.